File: cmd/hubctl/StepUp.go
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

package main

import (
	"context"
	"errors"
	"strings"

	"github.com/Jersyfi/hubtask/core/port/stepup"
	"github.com/Jersyfi/hubtask/presentation/openapi"
)

// The step-up, once, for every act that demands one (H-03, security.md §5).
//
// The server mints the demand in exactly one place - `stepup.Required`, "so that two operations
// cannot describe the same demand differently" - and this file is the client's half of that
// bargain: one place that recognises the refusal, asks the person, and repeats the call. A
// re-implementation per command is how `hubctl admin tenant delete` and `hubctl restore run` end
// up prompting differently for the same proof.
//
// What differs between the acts is only where the proof travels, and that is the caller's one
// line: `X-Hubtask-Step-Up` for the operations that take a header, and the request field for the
// restore, which has carried one since 0.4.5.

const stepUpPath = "/auth/step-up"

// proveAgain runs an act, and runs it once more with a fresh proof of identity if that is what the
// act was refused for.
//
// Once more, not in a loop: the proof is consumed by the one privileged action it is presented to,
// so a second refusal means the proof was rejected rather than missing, and asking a person for
// their password again on the same command would be a client arguing with a server.
func (cli *CLI) proveAgain(ctx context.Context, client *Client, act func(stepUp string) error) error {
	err := act("")
	methods, demanded := stepUpDemand(err)
	if !demanded {
		return err
	}

	token, err := cli.stepUp(ctx, client, methods)
	if err != nil {
		return err
	}
	return act(token)
}

// stepUpDemand reads the one refusal, and the methods it named. A 403 carrying any other code is
// somebody's answer to give, not this file's.
func stepUpDemand(err error) (string, bool) {
	var refusal APIError
	if !errors.As(err, &refusal) || refusal.DetailCode != stepup.CodeRequired {
		return "", false
	}
	return refusal.Params["methods"], true
}

// stepUp proves the person afresh and hands back the token the retry carries.
func (cli *CLI) stepUp(ctx context.Context, client *Client, methods string) (string, error) {
	// A step-up is a session's act - "a personal access token cannot prove a person afresh"
	// (auth.step_up_session_required). Said here rather than after the round trip, because the
	// answer is about this machine's configuration and a person who reads it has to sign in
	// rather than retype anything.
	if cli.Profile.Session.IsEmpty() {
		message, _ := cli.Catalogue.Message("auth.step_up_session_required", nil)
		return "", errorString(message)
	}

	request, err := cli.stepUpRequest(methods)
	if err != nil {
		return "", err
	}
	var grant openapi.StepUpGrant
	if err := client.Post(ctx, stepUpPath, request, &grant); err != nil {
		return "", err
	}
	// On standard error, like every diagnostic: what happened is worth seeing, and it is not part
	// of the answer the command was asked for.
	printf(cli.Err, "hubctl: proved again with the %s; retrying\n", strings.ToLower(string(grant.Method)))
	return grant.StepUpToken, nil
}

// stepUpRequest asks for one of the two proofs the contract accepts - "one of the two, never
// both".
//
// The authenticator's code wins where there is one to hand, because it is the stronger of the two
// and because a session under an armed factor is one whose password has already been typed today.
// Where there is none, the password is what is left, and the demand's own list is what says
// whether that is accepted.
func (cli *CLI) stepUpRequest(methods string) (openapi.StepUpRequest, error) {
	offersTOTP := methods == "" || strings.Contains(methods, "TOTP")
	if offersTOTP && (cli.Env(envTotp) != "" || !strings.Contains(methods, "PASSWORD")) {
		code, err := cli.readCredential(envTotp, "This needs proving again. The authenticator's current code: ")
		if err != nil {
			return openapi.StepUpRequest{}, err
		}
		return openapi.StepUpRequest{Code: &code}, nil
	}

	password, err := cli.readCredential(envPassword, "This needs proving again. Password (it will be visible): ")
	if err != nil {
		return openapi.StepUpRequest{}, err
	}
	return openapi.StepUpRequest{Password: &password}, nil
}

File: apps/webapp/src/lib/data/stepup.ts
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2026 Jérôme Bastian Winkel

/**
 * The step-up's two decisions, without the state that renders them (H-03).
 *
 * Here rather than in the store for this directory's usual reason: what is worth a test is the
 * shape of the recovery — try, ask once, retry once, never twice — and a rune cannot be run under
 * `node --test`. The store holds the pending prompt and hands `ask` to this.
 */

import { TransportError } from '@hubtask/sync-engine';

/** What the refusal said would be accepted. */
export type StepUpMethod = 'PASSWORD' | 'TOTP';

/**
 * Runs a call, and where the server refuses it for want of a proof, asks for one and runs it again.
 *
 * **Once.** A second `auth.step_up_required` after a fresh grant is the server saying the grant
 * was not what it wanted, and asking a third time would be a dialog somebody cannot escape. The
 * original refusal is what is thrown in that case, because it is the one that describes the
 * operation rather than the proof.
 *
 * A refusal the reader closes throws the refusal too: the operation did not happen, and saying so
 * with the server's own words is more use than a sentence this client invented about cancelling.
 */
export async function withStepUp<T>(
  call: (stepUpToken?: string) => Promise<T>,
  ask: (methods: readonly StepUpMethod[]) => Promise<string | undefined>,
): Promise<T> {
  try {
    return await call();
  } catch (cause) {
    if (!(cause instanceof TransportError) || !cause.needsStepUp) throw cause;

    const token = await ask(methodsOf(cause));
    if (token === undefined) throw cause;
    return call(token);
  }
}

/**
 * Which methods the refusal named, defaulting to the password.
 *
 * The server names them in the problem's parameters as a space-separated list — `PASSWORD TOTP`
 * for an account with a factor armed, `PASSWORD` for one without — and the contract says so at
 * `POST /auth/step-up`. Splitting on commas as well costs nothing and is what this once did
 * alone, which read the whole list as one unknown name and offered the password to exactly the
 * accounts a step-up protects (issue 544). A refusal that names none still has to produce a
 * usable prompt, and the password is the one every account has. A name this client does not
 * know is dropped rather than shown: a prompt with a field nobody can fill is worse than one
 * field fewer.
 */
export function methodsOf(cause: TransportError): readonly StepUpMethod[] {
  const named = cause.params?.methods;
  if (typeof named !== 'string' || named.trim() === '') return ['PASSWORD'];
  const methods = named
    .split(/[\s,]+/)
    .map((method) => method.trim().toUpperCase())
    .filter((method): method is StepUpMethod => method === 'PASSWORD' || method === 'TOTP');
  return methods.length > 0 ? methods : ['PASSWORD'];
}
